src/clj/fluree/db/query/exec/eval.cljc

```diff
@@ -210,6 +212,99 @@
   (let [dt-var (var->dt-var var)]
     `(iri/string->iri ~dt-var)))
 
+(def numeric-datatypes
+  #{const/iri-xsd-decimal
+    const/iri-xsd-double
+    const/iri-xsd-integer
+    const/iri-long
+    const/iri-xsd-int
+    const/iri-xsd-byte
+    const/iri-xsd-short
+    const/iri-xsd-float
+    const/iri-xsd-unsignedLong
+    const/iri-xsd-unsignedInt
+    const/iri-xsd-unsignedShort
+    const/iri-xsd-positiveInteger
+    const/iri-xsd-nonPositiveInteger
+    const/iri-xsd-negativeInteger
+    const/iri-xsd-nonNegativeInteger})
+
+(def string-datatypes
+  #{const/iri-string
+    const/iri-xsd-normalizedString
+    const/iri-lang-string
+    const/iri-xsd-token})
+
+(def comparable-datatypes
+  (set/union
+    numeric-datatypes
+    string-datatypes
+    #{const/iri-xsd-boolean
+      const/iri-anyURI
+      const/iri-id
+      const/iri-xsd-dateTime
+      const/iri-xsd-date
+      const/iri-xsd-time}))
+
+(defn infer-dt-iri
+  [x]
+  (get {const/$xsd:string  const/iri-string
```

Comment:
This is probably minor, but it wouldn't hurt to bind this map in a `def` so it doesn't have to be initialized and garbage collected with each invocation of this function.